src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will tie <see cref="Solution"/> and <see cref="RemoteHostClient.Session"/>'s lifetime together
+    /// so that one can handle those more easily
+    /// </summary>
+    internal sealed class SolutionAndSessionHolder : IDisposable
+    {
+        private readonly RemoteHostClient.Session _session;
+        private readonly PinnedRemotableDataScope _scope;
+        private readonly CancellationToken _cancellationToken;
+
+        public static async Task<SolutionAndSessionHolder> CreateAsync(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            var sessionWithSolution = new SolutionAndSessionHolder(session, scope, cancellationToken);
+
+            try
+            {
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return sessionWithSolution;
+            }
+            catch
+            {
+                sessionWithSolution.Dispose();
+
+                // we only expect this to happen on cancellation. otherwise, rethrow
+                cancellationToken.ThrowIfCancellationRequested();
+                throw;
+            }
+        }
+
+        private SolutionAndSessionHolder(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
```

Comment:
> Session is not tied to Solution.

I'm not asking for Session to be tied to Solution.  I'm saying that you can provide a *subclass* of Session that operates in the same manner as SolutionAndSessionHolder.  You can even call it something like:

```c#
class SessionWithSolution : Session {
      private readonly PinnedRemotableDataScope _scope;

      // overrides as appropriate
}
```

From the Feature perspective, all they have to do is say "...TryCreateSession(..., solution, ...)".  If they call this factory, they'll get one of these instances, but they'll only see it as a "Session".  They'll treat it just like they treat any other session they get today.  They call 'Invoke' helpers on it, and they 'Dispose' it when they're done.

There's no need for a feature to be aware that they got a *distinctly* different type here as they're just going to operate on all these in the exact same way.